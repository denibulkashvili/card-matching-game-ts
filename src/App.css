.App {
  text-align: center;
  width: 90%;
  display: block;
  margin: 0 auto;
}

.cards-section {
  display:grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 10px;
}
