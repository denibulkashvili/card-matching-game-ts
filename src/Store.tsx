import React, { createContext, useReducer } from "react";
import initialCards from "./cards/cards";
import { IAction, IState } from "./interfaces";

const initialState: IState = {
  cards: initialCards,
  isGameRunning: false,
  matchedCards: [],
  revealedCards: [],
};

export const Store = createContext<IState | any>(initialState);

const reducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case "TOGGLE_GAME_STATE":
      return { ...state, isGameRunning: action.payload };
    case "RESET_GAME":
      return { ...state, matchedCards: [], revealedCards: [], isGameRunning: action.payload };
    case "REVEAL_CARD":
      return { ...state, revealedCards: [...state.revealedCards, action.payload] };
    case "HANDLE_UNMATCHED_CARDS":
      return { ...state, revealedCards: [] };
    case "HANDLE_MATCHED_CARDS":
      return { ...state, matchedCards: [...state.matchedCards, ...action.payload], revealedCards: [] };
    default:
      return state;
  }
};

export const StoreProvider = (props: any): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return <Store.Provider value={{state, dispatch}}>{props.children}</Store.Provider>;
};
