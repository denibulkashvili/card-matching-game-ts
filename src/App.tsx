import React, { useContext } from "react";
import "./App.css";
import GameBoard from "./components/GameBoard";
import { IAction, IState } from "./interfaces";
import { Store } from "./Store";

const App: React.FC = () => {
  const { state, dispatch } = useContext(Store);

  const handleClick = (): IAction => {
    let dispatchObj = {
      payload: !state.isGameRunning,
      type: "TOGGLE_GAME_STATE",
    };
    if (state.isGameRunning) {
      dispatchObj = {
        payload: !state.isGameRunning,
        type: "RESET_GAME",
      };
    }
    return dispatch(dispatchObj);
  };

  const props: IState = {
    cards: state.cards,
    isGameRunning: state.isGameRunning,
    matchedCards: [],
    revealedCards: [],
  };

  return (
    <div className="App">
      {state.isGameRunning ? <GameBoard {...props} /> : <h3>Click "Start" to run the game</h3>}
      <div>
        <button onClick={handleClick} id="game-button">
          {state.isGameRunning ? <b>Stop</b> : <b>Start</b>}
        </button>
      </div>
    </div>
  );
};

export default App;
