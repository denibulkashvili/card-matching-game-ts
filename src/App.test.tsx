import React from "react";
import renderer from "react-test-renderer";
import App from "./App";

import { StoreProvider } from "./Store";

it("matches snapshot", () => {
  const component = renderer.create(
    <StoreProvider>
      <App />
    </StoreProvider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

