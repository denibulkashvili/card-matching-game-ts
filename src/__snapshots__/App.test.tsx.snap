// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="App"
>
  <h3>
    Click "Start" to run the game
  </h3>
  <div>
    <button
      onClick={[Function]}
    >
      <b>
        Start
      </b>
    </button>
  </div>
</div>
`;
